### PR TITLE
develop: Add -b/--build-directory option to set build_directory package attribute

### PR DESCRIPTION
**Pull request by vbrunini**
@becker33 we discussed this on slack a while back and I got around to making a PR for it.

### CODE REVIEW

**Comment by psakievich** [comment]
@spackbot fix style

**Comment by psakievich** [comment]
@spackbot fix style

**Comment by psakievich** [comment]
@spackbot fix style

**Comment by psakievich** [comment]
@spackbot fix style

File: lib/spack/spack/cmd/develop.py [hunk 1]
```diff
@@ -151,4 +153,17 @@ def develop(parser, args):
...
     with env.write_transaction():
+        if args.build_directory is not None:
+            scope = env.env_file_config_scope_name()
+            spack.config.add(
+                "packages:{}:package_attributes:build_directory:{}".format(
+                    spec.name, args.build_directory
+                ),
+                scope,
+            )
+            tty.warn(
```
Comment (psakievich) [inline lib/spack/spack/cmd/develop.py]: @becker33 I added this warning, but I don't think I like it. I imagine teams that rely on this feature will find it very annoying over time.  Do you think the doc's warning is sufficient?

Comment (becker33) [reply lib/spack/spack/cmd/develop.py]: I agree we should remove the warning

File: lib/spack/docs/environments.rst [hunk 2]
```diff
@@ -401,6 +401,18 @@ that git clone if ``foo`` is in the environment.
 Further development on ``foo`` can be tested by reinstalling the environment,
 and eventually committed and pushed to the upstream git repo.
 
+If the package being developed supports out-of-source builds then users can use the
```
Comment (becker33) [inline lib/spack/docs/environments.rst:404]: Can this be refactored to be:

1 paragraph explaining the feature

1 paragraph in a `.. warn:` block explaining the limitations.

That way the limitations will be more visible as a warning when the code warning is removed.

**Review by becker33 — CHANGES_REQUESTED** [review]

**Review by becker33 — APPROVED** [review]

**Pull request merged**